File: PA54.java
//Leetcode 160. Intersection of Two Linked Lists
//Time Complexity: O(n) // it is actually O(5n) where n is the no of nodes in the largest linked list
//Space Complexity: O(1) // No any extra space allocated
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA==null || headB==null){
            return null;
        }
        int lenA= 0;
        int lenB=0;
        ListNode ha= headA;
         ListNode hb= headB;
        while(ha!=null){
            lenA++;
            ha=ha.next;
        }
        while(hb!=null){
            lenB++;
            hb=hb.next;
        }
        
        while(lenA>lenB){
            headA=headA.next;
            lenA--;
        }
        while(lenB>lenA){
            headB=headB.next;
            lenB--;
        }
        
        while(headA!=headB){
            headB=headB.next;
            headA=headA.next;
        }
        return headA;
    }
}

File: PA_54.py
#LC 160 - Intersection of Two Linked Lists
#Time Complexity - O(len(A) + len(B))
#Space Complxity - O(1)
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        lA = 0
        lB = 0
        ta = headA
        tb = headB
        while ta:
            lA = lA + 1
            ta = ta.next
        while tb:
            lB = lB + 1
            tb = tb.next
        if lA > lB:
            while lA > lB:
                headA = headA.next
                lA = lA - 1
        else:
            while lA < lB:
                headB = headB.next
                lB = lB - 1
        while headA != headB:
            headA = headA.next
            headB = headB.next
        return headA